### ISSUE TITLE
[RFC] Project-level settings and new cwd setting

### ISSUE THREAD

**Issue by fabiomcosta** [chipsenkbeil/distant.nvim#93]
# Context

Today distant.nvim allows users to specify settings for all server and also to specific machine addresses.

 For example (see notes as comments on the example):
```lua
use {
  'chipsenkbeil/distant.nvim',
  config = function()
    require('distant').setup {
      -- This allows user to define their own defaults, which is applied to all servers
      ['*'] = {
        lsp =  {
          flow = {
            cmd = { '/default/path/to/rust-analyzer' },
            root_dir = '/default/path/to/project-rs'
          }
      },
      -- Settings for a specific server address
      ['unicorn.domain.com'] = {
        lsp =  {
          flow = {
            cmd = { '/unicorn/path/to/rust-analyzer' },
            root_dir = '/unicorn/path/to/project-rs'
          }
      } 
    }
  end
}
```

In this example when connecting to a distant-server on a machine whose address is `my.another.com`, the `lsp.flow` settings applied are going to be the default ones, but when connecting to a machine whose address is `unicorn.domain.com`, the ones specified for that specific server are going to take precedence over the default ones.

This works fine for the use case where a user has a single project on each of their machines, but it's very common to have more than one project on the same server. And currently there is no easy way to have different settings in the same server.

Another issue today, is that some lsps need to define a root_dir which defines where the lsp binary should be executed from (also known as current working directory or cwd), and the default cwd ends up being the cwd from the distant server, which is not always desired, and also always the same for a single server, which is not ideal for the multi-project server scenario that we are talking about.

To solve these I'm proposing the creation of project-level settings and the new cwd setting.

# Project-level settings

Project-level settings will allow users to easily change their settings based on a unique id that they define.
Once the project-level settings are defined, users are going to be able to switch between projects using a telescope-picker-like ui.

Here is a possible way to define these projects during distant.nvim's setup:

```lua
use {
  'chipsenkbeil/distant.nvim',
  config = function()
    require('distant').setup {
      -- This allows user to define their own defaults, which is applied to all servers
      ['*'] = {
        -- This can be used inside any machine-level settings, but using on the default only to keep the example small.
        projects = {
          rust = {
            -- you can define any of the existing settings here: https://github.com/chipsenkbeil/distant.nvim/blob/9dd21f8fa25795e56756e1ea27a1586ceee35582/lua/distant/settings.lua#L32-L62
            cwd = '/default/path/to/project',
            file = { mappings = { ... } },
          },
          portfolio = {
            cwd = '/path/to/portfolio/project',
          }
        },
        lsp =  {
            flow = {
              cmd = { '/default/path/to/rust-analyzer' },
              -- root_dir = this will default to the current project's cwd, do you'll likely not need to define it.
            }
         }
      },
    }
  end
}
```

Using these settings, distant will provide a new command (DistantProject maybe?) that will prompt the user to select the project by its name, using the a telescope-picker-like UI.
Selecting an option will apply the specified settings.

## Example project picker UI
![Screenshot 2023-04-16 at 12 27 06](https://user-images.githubusercontent.com/28530/232326602-c1509449-4fd5-45aa-827b-b1f8a66cbc47.png)

# Current working directly (CWD) setting

The new CWD setting works hand in hand with project-level settings.
On your local terminal, CWD is a global state that defines what directory you are currently executing commands from.
Think about this new setting as your remote (distant) CWD.

Any time you run a command through distant.nvim that ends up executing something on the server, the cwd that is on your setting will be used as the CWD on the server.

This is very useful not only for the `lsp` scenario that was already mentioned, but also when you have custom plugins that are spawning remote processes on the server you are connected through using distant.

**Comment by fabiomcosta** [comment]
I guess we don't necessarily need to provide the UI, it probably makes sense to be provided on a separate project.
In that case, distant.nvim would only provide a command that would allow that ui to work.
Something like: `:DistantProject project_id`

**Comment by chipsenkbeil** [comment]
> Another issue today, is that some lsps need to define a root_dir which defines where the lsp binary should be executed from (also known as current working directory or cwd), and the default cwd ends up being the cwd from the distant server, which is not always desired, and also always the same for a single server, which is not ideal for the multi-project server scenario that we are talking about.

Is this still the case? I thought that I updated the LSPs to set the current working directory to the specified `root_dir`: https://github.com/chipsenkbeil/distant.nvim/blob/9dd21f8fa25795e56756e1ea27a1586ceee35582/lua/distant/cli/client/lsp.lua#L56

> Any time you run a command through distant.nvim that ends up executing something on the server, the cwd that is on your setting will be used as the CWD on the server.

Sure. To make our lives easier, we'd probably want to consolidate the ways that someone can spawn a process in order to better control the default cwd used. Today, you have these options:

* `distant.fn.spawn` - for interactive use
* `distant.fn.spawn_wait` - uses `distant.fn.spawn` underneath, so anything we do for that will apply here
* `wrap({ cmd = '...' })` - as well as the lsp & shell variants
* `:DistantRun` - uses `distant.fn.spawn_wait` underneath, so anything we do for that will apply here
* Internal CLI builders for `action`, `lsp`, and `shell` - I'd say leave these alone as they are the base way to build the `wrap` commands.

> I guess we don't necessarily need to provide the UI, it probably makes sense to be provided on a separate project.

Once I merge in #89, I'm actually a big fan of incorporating as much UI-friendly exposure as possible into the plugin itself. I just don't want to depend on a 3rd party plugin like telescope given that the distant.nvim plugin requires no additional lua plugins (plenary, telescope, etc) to get the base functionality.

If you want to start by providing an extension with telescope, I don't mind that being added to the extension we offer for searching today. Feel free to add that functionality in the `search.lua` file alongside the search picker: https://github.com/chipsenkbeil/distant.nvim/blob/9dd21f8fa25795e56756e1ea27a1586ceee35582/lua/telescope/_extensions/distant/search.lua#L29

**Comment by fabiomcosta** [comment]
> Is this still the case? I thought that I updated the LSPs to set the current working directory to the specified root_dir:

The problem is that the CWD of the LSP in this case is per machine and not per project.
If you have 2 different project roots that are using an LSP, like JS projects using flow for example, there is no easy way to switch between the 2 projects/cwds today.

**Comment by fabiomcosta** [comment]
Some observations while trying to implement shit RFC:

* ATM, project-level settings only make sense to be used to redefine the CWD, and LSP (in some edge cases where the binary might be different per-project).
* The project is only defined after connecting to the server, so the server settings are still used during the connection no matter what, and then the project settings can be applied. I can see this being confusing to users.

I think that I'm starting to conclude that "project" should be a setting, but not be able to redefine any of the server settings.
You can define your projects per server setting (see example).

Here are some assertions for what `project` would be:
* The `project` would live inside the global state along `client`, `manager` and others.
* `project` for now would only contain a `cwd` (final name TBD), but we can extend it in the future in case we find more project specific properties.
* After connection, the user would be by default using the "server" project, which is a default using the server's cwd.
* Once a project is selected by the user using the `:DistantProject` (final name TBD) the `project` cwd value would be used as the `root_dir` for any LSP that is started, and also as the current-dir property for any distant client call, as well as used when calling `distant.wrap`.

Example of the projects setting:
```lua
use {
  'chipsenkbeil/distant.nvim',
  config = function()
    require('distant').setup {
      -- This allows user to define their own defaults, which is applied to all servers
      ['*'] = {
        projects = {
          rust = { cwd = '/path/to/rust-project' },
          portfolio = { cwd = '/path/to/portfolio/project' }
        },
        // ...
      },
      ['unicorn.domain.com'] = {
        projects = {
          other_project = { cwd = '/path/to/other-project' },
        },
        // ...
      },
    }
  end
}
```

**Comment by fabiomcosta** [comment]
Iterating more on this.

The "project" concept makes it so letting the user choose their own cwd always would require them to select a "project", and that might not always be the case. This might just be complicating things.

I'm thinking that we should just have the `cwd` setting, and that can either be a string, or a map of cwd aliases/ids that will facilitate changing between them by using the alias instead of having to type the full path.

```lua
use {
  'chipsenkbeil/distant.nvim',
  config = function()
    require('distant').setup {
      ['*'] = {
        cwd = {
          rust = '/path/to/rust-project',
          portfolio = '/path/to/portfolio/project'
        },
        -- ...
      },
      ['horse.domain.com'] = {
        cwd = {
          personal_website = '/path/to/personal_website/project',
        },
        -- ...
      },
      ['unicorn.domain.com'] = {
        cwd = '/path/to/unicorn-project',
        -- ...
      },
    }
  end
}
```

If we consider this change for this RCF, here are some assertions of this RFC:
* The `cwd` would live inside the global state along `client`, `manager` and others.
* the `state` instance would have a new `get_cwd` method which would return the currently selected `cwd` as a string.
* If `cwd` is not set  by the user, `get_cwd` returns nil.
* Once a `cwd` is selected by the user using the `:DistantCwd` (final name TBD) the `cwd` value would be used as the `root_dir` for any LSP that is started, and also as the `current_dir` property for any distant client call, as well as used when calling `distant.wrap`.

**Comment by chipsenkbeil** [comment]
@fabiomcosta I haven't made a tag yet for the v0.2 branch, so I'm thinking of reworking the setup function a little bit before tagging a stable moment for v0.2 of the plugin.

One challenge is that settings are a bit complex because you're providing them all relative to one or more servers. So my first thought is to separate out server settings into a specific field like this:

```lua
require('distant').setup {
    servers = {
        ['*'] = {
            cwd = {
              rust = '/path/to/rust-project',
              portfolio = '/path/to/portfolio/project'
            },
            -- ...
        },
        ['horse.domain.com'] = {
            cwd = {
              personal_website = '/path/to/personal_website/project',
            },
            -- ...
        },
        ['unicorn.domain.com'] = {
            cwd = '/path/to/unicorn-project',
            -- ...
        },
    }
}
```

But this means that we could have a separate setting for project-specific details that would make them separate from a specific server:

```lua
require('distant').setup {
    servers = { ... },
    projects = {
        portfolio = { ... },
        personal_website = { ... },
    },
}
```

Not sure exactly what we want in each nor how we'd handle merging them together (yet). The reason I want to move server settings to a sub-table is so I can have agnostic settings that don't require the clunky means of stuffing everything in to "*".

```lua
require('distant').setup {
    -- Server-specific settings
    servers = { ... },

    -- Project-specific settings
    projects = { ... },

    -- General settings
    network = { private = true },
    client { bin = '/path/to/distant', log_file = '/path/to/log.txt' },
}
```

**Comment by chipsenkbeil** [comment]
@fabiomcosta can you take a look at the upcoming settings here?

https://github.com/chipsenkbeil/distant.nvim/blob/670cb2ac3ac373cc01e556f24452c476c3f65ed2/lua/distant/default.lua

I'm testing these out in the overhaul interface branch.  I don't have `cwd` set yet for servers like you've shown above, and haven't added anything for projects yet. We can tackle this issue once I merge that branch, but wanted to give you a visual as to what the current setup will look like.

```lua
require('distant').setup {
    servers = {
        ['server.example.com'] = {
            cwd = '...',
        }
    }
}
```

If you wanted to be able to work with projects that change settings, I was thinking of having another tab in the UI for projects and let you pick from the list there.

**Comment by chipsenkbeil** [comment]
@fabiomcosta I added this to v0.3.0 milestone as I'm going to merge my overhaul UI and rewrite work into a new version of the plugin, given that it changes the setup function and other APIs. We can incorporate your project RFC into v0.3 post-rewrite.

Rewrite tracking issue: #96.

**Comment by chipsenkbeil** [comment]
@fabiomcosta pretty much everything I want available for v0.3 of this plugin is done. I'm going to move this to be something we can introduce post-release of v0.3.

The distant v0.20 release is just about done, which is what I want (versus an alpha) for v0.3 to switch to the main release. Documentation is really the last hurdle before this is good to go.

See [Distant v0.20.0 milestone](https://github.com/chipsenkbeil/distant/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.20).

**Comment by fabiomcosta** [comment]
I believe this task has been addressed at this point
